File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.length();
        vector<int> hash(256,-1);

        int maxLength =0;
        int l=0;
        int r =0;

        while(r<n){
            if(hash[s[r]]!=-1){
                if(hash[s[r]]>=l){
                    l = hash[s[r]]+1;
                }
            }

            hash[s[r]]=r;
            maxLength = max(maxLength,r-l+1);
            r++;
        }
        return maxLength;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        n = len(s)
        last = [-1]*256
        l=0
        r=0
        maxLength =0

        while r<n:
            index = ord(s[r])
            if last[index]!=-1:
                if last[index]>=l:
                    l=last[index]+1
            
            last[index]=r
            maxLength = max(maxLength,r-l+1)
            r+=1
        
        return maxLength